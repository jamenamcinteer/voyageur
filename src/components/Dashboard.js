import React, { useState } from "react";
import { Link } from "react-router-dom";
import Header from "./Navigation/Header";
import ButtonLink from "./Buttons/ButtonLink";
import Button from "./Buttons/Button";
import TripCard from "./Cards/TripCard";
import moment from "moment";
import Modal from "react-modal";
import styled from "styled-components";
import {} from "twix";
import { connect } from "react-redux";
import { CardContainer } from "./StyledComponents/Layout";

const ModalHeader = styled.h3`
  font-size: 1em;
  font-weight: normal;
  margin-top: 0;
`;

const ModalLink = styled(Link)`
  display: block;
  padding: 10px 20px;
  border-top: 1px solid ${props => props.theme.themeColorSecondary};
  text-decoration: none;
  margin-left: -20px;
  margin-right: -20px;
  color: ${props => props.theme.darkFont};

  &:last-child {
    padding-bottom: 0;
  }
`;

const ButtonContainer = styled.div`
  display: flex;
  justify-content: space-between;
  margin: 20px 10px;

  @media (min-width: 1024px) {
    justify-content: center;
  }

  button,
  a {
    max-width: 200px;
  }
`;

const OneButtonBackground = styled.div`
  height: calc(100vh - 69px);
  width: 100vw;
  padding: 70px;
  display: flex;
  align-items: center;
  text-align: center;
  position: relative;

  &:before {
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-image: url("/dashboard-background.jpg");
    background-size: cover;
    background-position: top right;
    filter: blur(1px) brightness(40%) grayscale(50%) sepia(10%);
  }
`;

const OneButtonContainer = styled.div`
  position: relative;
  width: 100%;
`;

const DashboardHeader = styled.h4``;

const Dashboard = props => {
  const [addExpenseModal, setAddExpenseModal] = useState(false);

  const today = new Date();

  let futureTrips = [];
  let pastTrips = [];
  // get all trips in the future from "closest" to "farthest"
  props.trips &&
    props.trips.map(trip => {
      if (trip.endDate >= moment(today).valueOf()) {
        futureTrips.push(trip);
      }
      return true;
    });
  futureTrips.sort((a, b) => parseInt(a.endDate) - parseInt(b.endDate));
  // get all trips in the past from "closest" to "farthest"
  props.trips &&
    props.trips.map(trip => {
      if (trip.endDate < moment(today).valueOf()) {
        pastTrips.push(trip);
      }
      return true;
    });
  pastTrips.sort((a, b) => parseInt(b.endDate) - parseInt(a.endDate));

  const getDates = (startDate, endDate) => {
    return moment(startDate)
      .twix(endDate, { allDay: true })
      .format({ monthFormat: "MMM", dayFormat: "D" });
  };

  return (
    <React.Fragment>
      <Header
        title="My Trips"
        theme={props.theme}
        auth={props.auth}
        backHide={true}
      />
      {props.trips.length > 0 && (
        <React.Fragment>
          <ButtonContainer>
            <ButtonLink
              to="/trip/add"
              buttonText="Add a Trip"
              buttonType="secondary"
              buttonWidth="50%"
              theme={props.theme}
            />
            <Button
              handleClick={e => setAddExpenseModal(true)}
              buttonText="Add Expense"
              buttonType="primary"
              buttonWidth="50%"
              theme={props.theme}
            />
          </ButtonContainer>
          <Modal
            isOpen={addExpenseModal}
            onRequestClose={e => setAddExpenseModal(false)}
            contentLabel="Add Expense Modal"
            ariaHideApp={props.isTest ? false : true}
            style={{
              overlay: {
                backgroundColor: "rgba(0,0,0,.5)"
              },
              content: {
                borderRadius: "0",
                // minHeight: "20vh",
                maxHeight: "40vh",
                overflowY: "auto",
                width: "80vw",
                top: "30vh",
                left: "10vw",
                right: "auto",
                bottom: "auto"
              }
            }}
          >
            <ModalHeader>Choose a Trip:</ModalHeader>
            {futureTrips.length > 0 &&
              futureTrips.map((trip, index) => {
                return (
                  <ModalLink
                    to={`/trip-budget/${trip._id}/add-expense`}
                    key={trip._id}
                  >
                    {trip.destination} ({getDates(trip.startDate, trip.endDate)}
                    )
                  </ModalLink>
                );
              })}
            <Button
              handleClick={e => setAddExpenseModal(false)}
              buttonText="Cancel"
              buttonType="link"
              customStyles={{
                background: { padding: "10px 0", margin: "0", float: "right" }
              }}
              dataTestid="closeModal"
            />
          </Modal>
          <CardContainer>
            {/* <h4>Upcoming Trips</h4> */}
            {futureTrips.length > 0 &&
              futureTrips.map(trip => {
                const budgetItems = props.budgetItems.filter(
                  bItem => bItem.tripId === trip._id
                );
                const expenses = props.expenses.filter(
                  expense => expense.tripId === trip._id
                );
                return (
                  <TripCard
                    key={trip._id}
                    destination={trip.destination}
                    photo={trip.photo}
                    photoAttribution={trip.photoAttribution}
                    to={`/trip/${trip._id}`}
                    startDate={trip.startDate}
                    endDate={trip.endDate}
                    budgetItems={budgetItems}
                    expenses={expenses}
                    theme={props.theme}
                  />
                );
              })}
            {pastTrips.length > 0 && (
              <DashboardHeader>Past Trips</DashboardHeader>
            )}
            {pastTrips.length > 0 &&
              pastTrips.map(trip => {
                const budgetItems = props.budgetItems.filter(
                  bItem => bItem.tripId === trip._id
                );
                const expenses = props.expenses.filter(
                  expense => expense.tripId === trip._id
                );
                return (
                  <TripCard
                    key={trip._id}
                    destination={trip.destination}
                    photo={trip.photo}
                    photoAttribution={trip.photoAttribution}
                    to={`/trip/${trip._id}`}
                    startDate={trip.startDate}
                    endDate={trip.endDate}
                    budgetItems={budgetItems}
                    expenses={expenses}
                    theme={props.theme}
                  />
                );
              })}
          </CardContainer>
        </React.Fragment>
      )}
      {props.trips.length === 0 && (
        <OneButtonBackground>
          <OneButtonContainer>
            <ButtonLink
              to="/trip/add"
              buttonText="Add a Trip"
              buttonType="primary"
              customStyles={{
                text: { fontSize: "1.5em" }
              }}
            />
          </OneButtonContainer>
        </OneButtonBackground>
      )}
    </React.Fragment>
  );
};

const mapStateToProps = state => {
  return {
    trips: state.trips,
    budgetItems: state.budgetItems,
    expenses: state.expenses,
    auth: state.auth
  };
};

export default connect(mapStateToProps)(Dashboard);
