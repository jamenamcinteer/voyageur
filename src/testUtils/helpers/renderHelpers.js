import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../../theme";
import "jest-styled-components";
import { createStore } from "redux";
import { Provider } from "react-redux";
import tripsReducer, { tripsReducerDefaultState } from "../../reducers/trips";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

export function renderWithTheme(component) {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
}

export function renderWithReduxRouterAndTheme(
  ui,
  {
    initialState,
    store = createStore(tripsReducer, tripsReducerDefaultState),
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) {
  return {
    ...render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <Router history={history}>{ui}</Router>
        </ThemeProvider>
      </Provider>
    ),
    store,
    history
  };
}
