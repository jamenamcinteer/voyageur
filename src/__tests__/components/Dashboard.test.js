import React from "react";
import { cleanup } from "@testing-library/react";
import { createStore } from "redux";
import Dashboard from "../../components/Dashboard";
import {
  trips,
  budgetCategories,
  budgetItems,
  expenses,
  auth
} from "../../testUtils/fixtures/mockStoreData";
import { renderWithReduxRouterAndTheme } from "../../testUtils/helpers/renderHelpers";
import { toBeInTheDocument } from "@testing-library/jest-dom";

expect.extend({ toBeInTheDocument });

afterEach(cleanup);

const store = createStore(() => ({
  trips,
  budgetCategories,
  budgetItems,
  expenses,
  auth
}));

const historyMock = { push: jest.fn() };
// Object.defineProperty(window, 'innerWidth', {writable: true, configurable: true, value: 200})
window.innerWidth = 1024

test("can render with redux with custom store", () => {
  const { asFragment } = renderWithReduxRouterAndTheme(<Dashboard />, {
    store
  });
  expect(asFragment()).toMatchSnapshot();
});

test("should display modal after clicking Add Expense with list of trips to choose from", () => {
  const { getByText, getByTestId, queryByText } = renderWithReduxRouterAndTheme(
    <Dashboard history={historyMock} isTest={true} />,
    { store }
  );

  getByText("Add Expense").click();
  expect(getByText("Choose a Trip:")).toBeInTheDocument();
  expect(getByText("Hawaii (Apr 26, 1970 - Oct 1, 8307)")).toBeInTheDocument();

  getByTestId("closeModal").click();
  expect(queryByText("Choose a Trip:")).toBeNull();
});
