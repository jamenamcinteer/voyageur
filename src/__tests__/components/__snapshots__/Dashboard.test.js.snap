// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render with redux with custom store 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #dfe7ed;
}

.c1 {
  width: 100%;
  max-width: 960px;
  margin: 0 auto;
  padding: 10px 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  font-size: 1.2em;
  font-weight: normal;
  color: #4b5c6b;
}

.c3 {
  background: 0;
  border: 0;
  padding: 0;
  border: 3px solid #c7d2db;
  border-radius: 100px;
}

.c3:focus {
  outline: 0;
  border: 3px solid #A63E26;
}

.c4 {
  border-radius: 100px;
  width: 45px;
  height: 45px;
  display: block;
}

<header
    class="c0"
  >
    <div
      class="c1"
    >
      <h1
        class="c2"
      >
        My Trips
      </h1>
      <button
        class="c3"
      >
        <img
          alt="Account Options"
          class="c4"
          src="https://picsum.photos/id/947/45/45"
        />
      </button>
    </div>
  </header>
  .c2 {
  background-color: #fff;
  border: 1px solid #0C7C88;
  font-size: 1em;
  padding: 10px 20px;
  border-radius: 5px;
  width: 50%;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
  text-align: center;
  margin: 10px;
  position: relative;
  overflow: hidden;
  -webkit-transition: background-color 0.3s;
  transition: background-color 0.3s;
}

.c2 > * {
  position: relative;
}

.c2:before {
  content: "";
  position: absolute;
  top: 50%;
  left: 50%;
  display: block;
  width: 0;
  padding-top: 0;
  border-radius: 100%;
  background-color: rgba(236,240,241,0.3);
  -webkit-transform: translate(-50%,-50%);
  -moz-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  -o-transform: translate(-50%,-50%);
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c2:active:before {
  width: 120%;
  padding-top: 120%;
  -webkit-transition: width 0.2s ease-out,padding-top 0.2s ease-out;
  transition: width 0.2s ease-out,padding-top 0.2s ease-out;
}

.c2:active {
  -webkit-transform: translateY(2px);
  -ms-transform: translateY(2px);
  transform: translateY(2px);
}

.c2:focus {
  outline: 3px solid #A63E26;
}

.c3 {
  font-size: 1em;
  font-weight: normal;
  color: #0C7C88;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 {
  background-color: #0C7C88;
  border: 0;
  padding: 10px 20px;
  border-radius: 5px;
  font-size: 1em;
  width: 50%;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
  text-align: center;
  margin: 10px;
  position: relative;
  overflow: hidden;
  -webkit-transition: background-color 0.3s;
  transition: background-color 0.3s;
}

.c4 > * {
  position: relative;
}

.c4:before {
  content: "";
  position: absolute;
  top: 50%;
  left: 50%;
  display: block;
  width: 0;
  padding-top: 0;
  border-radius: 100%;
  background-color: rgba(236,240,241,0.3);
  -webkit-transform: translate(-50%,-50%);
  -moz-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  -o-transform: translate(-50%,-50%);
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c4:active:before {
  width: 120%;
  padding-top: 120%;
  -webkit-transition: width 0.2s ease-out,padding-top 0.2s ease-out;
  transition: width 0.2s ease-out,padding-top 0.2s ease-out;
}

.c4:active {
  -webkit-transform: translateY(2px);
  -ms-transform: translateY(2px);
  transform: translateY(2px);
}

.c4:focus {
  outline: 3px solid #A63E26;
}

.c5 {
  font-size: 1em;
  font-weight: normal;
  color: #fff;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c13 {
  width: 100%;
  height: 10px;
  position: relative;
  background: #dce0e3;
  border-radius: 5px;
}

.c14 {
  display: block;
  height: 100%;
  border-radius: 5px;
  position: relative;
  overflow: hidden;
}

.c10 {
  position: absolute;
  bottom: 0;
  background-color: #333;
  color: #fff;
  padding: 5px;
  width: 100%;
  font-size: 0.75em;
}

.c10 a {
  color: #fff;
}

.c7 {
  background-color: #fff;
  border: 1px solid #c7d2db;
  padding: 20px;
}

.c11 {
  font-size: 1.2em;
  font-weight: normal;
  color: #4b5c6b;
}

.c12 {
  font-size: 1em;
  font-weight: normal;
  color: #4b5c6b;
}

.c8 {
  width: calc(100% + 40px);
  margin: -20px -20px 0 -20px;
  position: relative;
}

.c9 {
  width: 100%;
  height: 200px;
  display: block;
}

.c15 {
  margin: 20px;
}

.c6 {
  margin: 20px;
  display: grid;
  grid-gap: 20px;
  grid-template-columns: repeat(auto-fill,minmax(280px,1fr));
}

.c0 {
  width: 100%;
  max-width: 960px;
  margin: 0 auto;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 20px 10px;
}

.c1 button,
.c1 a {
  max-width: 200px;
}

@media (min-width:1024px) {
  .c7 {
    display: grid;
    grid-template-columns: 280px 1fr;
    grid-gap: 40px;
    padding-bottom: 0;
  }
}

@media (min-width:1024px) {
  .c6 {
    display: block;
  }
}

<section
    class="c0"
  >
    <div
      class="c1"
    >
      <a
        backgroundcolor="#fff"
        border="1px solid #0C7C88"
        buttontype="secondary"
        buttonwidth="50%"
        class="c2"
        href="/trip/add"
      >
        <span
          class="c3"
          color="#0C7C88"
        >
          Add a Trip
        </span>
      </a>
      <button
        class="c4"
      >
        <span
          class="c5"
          color="#fff"
        >
          Add Expense
        </span>
      </button>
    </div>
    <div
      class="c6"
    >
      <a
        href="/trip/1"
        style="text-decoration: none;"
      >
        <div
          class="c7"
        >
          <div
            class="c8"
          >
            <img
              alt=""
              class="c9"
              src="https://images.unsplash.com/photo-1549317487-92cec5e76645?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjc5ODc0fQ&w=333&h=200&fit=crop&crop=focalpoint"
            />
            <figcaption
              class="c10"
            >
              photo attribution
            </figcaption>
          </div>
          <div>
            <h1
              class="c11"
            >
              Hawaii
            </h1>
            <h2
              class="c12"
            >
              Apr 26, 1970 - Oct 1, 8307
            </h2>
            <div
              class="c13"
            >
              <span
                class="c14"
                style="width: 29.731%; background-color: rgb(12, 124, 136);"
              />
            </div>
          </div>
        </div>
      </a>
    </div>
    <div
      class="c15"
    />
    <div
      class="c6"
    />
  </section>
</DocumentFragment>
`;
